File: arl2arw.f90
program arl2arw
! -----------------------------------------------------------------------------
!
! Program to convert the packed ARL meteorological file format to WRF netCDF
! More information for ARL packed metorological format: 
! - https://ready.arl.noaa.gov/hysplitusersgudaye/S141.htm
! - https://www.ready.noaa.gov/archives.php
! - ftp://arlftp.arlhq.noaa.gov/pub/archives/utility/chk_data.f
!
! -----------------------------------------------------------------------------
    use netcdf
    implicit none
! -----------------------------------------------------------------------------
! Unpacking

    real, allocatable :: real_data(:, :) ! 2d unpacked data 
    character(1), allocatable :: char_data(:) ! 1d packed data

    character(80) :: arl_file ! prompt file name
    character(50) :: label ! header label 
    character(3072) :: header ! file header

    integer :: year, month ! year and month
    integer :: day, hour ! day and hour
    character(4) :: var_desc ! descriptor of variable being written
    integer :: ifc ! No clue 
    integer :: level ! level of z 
    integer :: k, n ! place holders
    integer :: nexp ! integer scaling exp of packaed data array
    real :: var1 ! real value of array at (1, 1)
    real :: prec ! real precision of packed data array

    character(4) :: model ! data model character
    integer :: grid_num ! grid number
    integer :: z_coord ! z coordinate
    real :: pole_lat, pole_lon, ref_lat, ref_lon, grid_size, orient, & 
            tan_lat, sync_xp, sync_yp, sync_lat, sync_lon, reserved
    integer :: nx, ny, nz, nxy ! number of x, y points, z levels
    integer :: recl ! record length
    
    integer :: alstat ! /= 0 if not enough memory 
    integer :: iostat ! > 0 if not enough memory 

! -----------------------------------------------------------------------------
! netCDF
    
    character(*), parameter :: ncpath = 'output.nc' 
    integer :: ncid 

    character(80) :: nc_file
    real, allocatable :: coord_data(:, :)

    integer :: lon_varid, lat_varid, rec_varid

    ! writing 4-d data lat-long grid with 
    character(*), parameter :: name_lat = 'latitude'
    character(*), parameter :: name_lon = 'longitude'
    character(*), parameter :: name_level = 'level'

    integer :: lvl_dimid, lon_dimid, lat_dimid, rec_dimid

    character(*), parameter :: units_lat = 'degrees_north'
    character(*), parameter :: units_lon = 'degrees_east'

    integer :: nx_id, ny_id, nz_id

    integer :: nlvls, nlats, nlons

    integer :: start(4)

    real, parameter :: start_lat = 25.0, start_lon = -119.0

    real :: lats(10), lons(10)
    integer :: lon, lat

    character(19) :: records 

! -----------------------------------------------------------------------------

    interface

        ! unpacked character data is resolved to real_data 
        subroutine unpack(char_data, real_data, nx, ny, nexp, var1)
            character(1), intent(in) :: char_data(:)
            real, intent(out) :: real_data(:, :)
            integer, intent(in) :: nx, ny, nexp
            real, intent(in) :: var1
        end subroutine

        ! wrapping netCDF calls with "call check()" checks the error status
        subroutine check(status)
            integer, intent(in) :: status
        end subroutine

    end interface

! -----------------------------------------------------------------------------
    
    ! get file name
    write(*,*) 'Enter ARL file name:'
    read(*, '(a)') arl_file
    arl_file = adjustl(arl_file)

    ! open the file and decode the 50-byte header plus the fixed portion
    open(10, file = arl_file, recl = 158, access = 'direct', form = 'unformatted')

    ! decode the "standard" portion of the index record
    read(10, rec = 1) label, header(1:108)
    read(label,'(5i2,4x,a4)') year, month, day, hour, ifc, var_desc
    write(*, '(a, 4i2)') 'YYMMDDHH: ', year, month, day, hour
    
    ! decode extended portion of header
    ! field descriptions: https://ready.arl.noaa.gov/hysplitusersguide/S141.htm
    read(header(1:108), '(a4, 1i3, 1i2,  12f7.0, 3i3)')         & 
         model, grid_num, z_coord,                              &
         pole_lat, pole_lon, ref_lat,                           &
         ref_lon, grid_size, orient,                            &
         tan_lat, sync_xp, sync_yp,                             &
         sync_lat, sync_lon, reserved,                          &
         nx, ny, nz  
    
    ! close the header read
    close(10)

    nxy = nx * ny ! record size
    recl = nxy + 50 ! record size + header length

    ! properly re-open the file with the record length
    open(10, file = arl_file, recl = recl, access = 'direct', form = 'unformatted')

    print *, 'Model: ', model, new_line('a'),                           & 
             'Grid Number: ', grid_num, new_line('a'),                  & 
             'Z-Coordinate: ', z_coord, new_line('a'),                  & 
             'Pole (lat,lon): ', pole_lat, pole_lon, new_line('a'),     & 
             'Reference (lat,lon): ', ref_lat, ref_lon, new_line('a'),  & 
             'Reference grid size: ', grid_size, new_line('a'),         &
             'Orientation: ', orient, new_line('a'),                    & 
             'Tangent Latitude: ', tan_lat, new_line('a'),              & 
             'Synch (x,y): ', sync_xp, sync_yp, new_line('a'),          & 
             'Synch (lat,lon): ', sync_lat, sync_lon, new_line('a'),    & 
             'Reserved: ', reserved, new_line('a'),                     &
             'Number of X points: ', nx, new_line('a'),                 & 
             'Number of Y points: ', ny, new_line('a'),                 & 
             'Number of Levels: ', nz, new_line('a')

    ! allocate the character and real arrarys of proper dimensions
    allocate(real_data(nx,ny), stat = alstat)
    allocate(char_data(nxy), stat = alstat)

    ! read the file and upack its content     
    n = 1 
    do
        ! read each header label
        read(10, rec = n, iostat = iostat) label, (char_data(k), k = 1, nxy) 
        if (iostat > 0) then ! >0 is error is thrown / end of records(?)
            exit 
        else
            read(label, '(6i2, 2x, a4, 1i4, 2f7.0)') & 
            year, month, day, hour, grid_num, level, & 
            var_desc, nexp, prec, var1
            ! unpack data from header 
            if(var_desc /= 'INDX')  then 
                call unpack(char_data, real_data, nx, ny, nexp, var1)   
            end if
            ! preview a bit first row of each data
            ! print *, level, var_desc, var1, real_data(2:3, 1), '...'
        end if
        n = n + 1
        ! TODO: implement netcdf writeout
    end do

    ! -------------------------------------------------------------------------

    write(*,*) 'Enter a NetCDF file name:'
    read(*, '(a)') nc_file
    nc_file = adjustl(nc_file)

    call check(nf90_open(nc_file, nf90_nowrite, ncid))

    ! Get the varids of the latitude and longitude coordinate variables.
    call check(nf90_inq_varid(ncid, 'XLAT', lat_varid))
    call check(nf90_inq_varid(ncid, 'XLON', lon_varid))

      ! Read the latitude and longitude data.
    call check( nf90_get_var(ncid, lat_varid, lats) )
    call check( nf90_get_var(ncid, lon_varid, lons) )

    ! print *, lats

    do rec = 1, 480 ! TODO dynamic allocate
        start(4) = rec
        call check(nf90_get_var(ncid, lat_varid, count = rec - 1))
        call check(nf90_get_var(ncid, lon_varid, count = rec - 1))

        i = 0
        do lvl = 1, nlvls
            do lat = 1, nlats
                do lon = 1, nlons
                    ! Write each level variable
                    i = i + 1
                end do
            end do
        end do
     ! next record
    end do
    ! -----------------------------------------------------------------------
    ! Header dump previews of ARL and WRF-netCDF

    !    wrf.arl
    !YYMMDDHH: 20112521
    !Model: AWRF
    !Grid Number:           21 
    !Z-Coordinate:            0 
    !Pole (lat,lon):    45.6899986      -121.000000     
    !Reference (lat,lon):    45.6899986      -121.000000     
    !Orientation:    0.00000000     
    !Tangent Latitude:    45.6899986     
    !Synch (x,y):    203.000000       141.500000     
    !Synch (lat,lon):    45.3499985      -118.930000     
    !Reserved:    0.00000000     
    !Number of X points:          405 
    !Number of Y points:          282 
    !Number of Levels:           38 

    !   wrf.nc 
    !$ ncdump -h wrfout_d3.2020112500.f21.0000
    !netcdf wrfout_d3.2020112500.f21 {
    !dimensions:
    !   Time = UNLIMITED ; // (1 currently)
    !   DateStrLen = 19 ;
    !   west_east = 405 ;
    !   south_north = 282 ;
    !   bottom_top = 37 ;
    !   bottom_top_stag = 38 ;
    !   soil_layers_stag = 4 ;
    !   west_east_stag = 406 ;
    !   south_north_stag = 283 ;
    !   seed_dim_stag = 2 ;
    ! -------------------------------------------------------------------------

    ! Create the nc file
    call check(nf90_create(ncpath, nf90_clobber, ncid))

    ! nc dimensions
    call check(nf90_def_dim(ncid, 'Time', nf90_unlimited, rec_dimid))
    call check(nf90_def_dim(ncid, 'west_east', nx, nx_id))
    call check(nf90_def_dim(ncid, 'south_north', ny, ny_id))
    call check(nf90_def_dim(ncid, 'bottom_top', nz - 1, nz_id))
    call check(nf90_def_dim(ncid, 'bottom_top_stag', nz, nz_id))

    ! nc variables
    call check(nf90_def_var(ncid, 'Time', nf90_char, rec_dimid, rec_varid))
    call check(nf90_def_var(ncid, 'XLAT', nf90_real, ny_id, lat_varid))
    call check(nf90_def_var(ncid, 'XLONG', nf90_real, nx_id, lon_varid))

    !call check(nf90_def_var(ncid, ''))

    ! nc attributes
    call check(nf90_put_att(ncid, lat_varid, 'description', 'LATITUDE, SOUTH IS NEGATIVE'))
    call check(nf90_put_att(ncid, lat_varid, 'units', 'degree_north'))
    call check(nf90_put_att(ncid, lon_varid, 'description', 'LONGITUDE, WEST IS NEGATIVE'))
    call check(nf90_put_att(ncid, lon_varid, 'units', 'degree_east'))

    ! end nc define 
    call check(nf90_enddef(ncid))

    ! Messing about
    ! do lat = 1, 10 
    !     lats(lat) = start_lat + (lat - 1) * 5.0
    ! end do
    ! do lon = 1, 10 
    !     lons(lon) = start_lon + (lon - 1) * 5.0
    ! end do

    ! !call check(nf90_out_var(ncid, rec_varid, records))
    ! call check(nf90_put_var(ncid, lon_varid, lons))
    ! call check(nf90_put_var(ncid, lat_varid, lats))

    ! close the nc file
    call check(nf90_close(ncid))
    
end program arl2arw

! -----------------------------------------------------------------------------
! Subroutines 
subroutine unpack(char_data, real_data, nx, ny, nexp, var1) 

    character(1), intent(in) :: char_data(:)    
    real, intent(out) :: real_data(:, :)
    integer, intent(in) :: nx, ny, nexp
    real, intent(in) :: var1
    real :: scale
    real :: vold
    integer :: indx
    integer :: i, j

    scale = 2.0 ** (7 - nexp)
    vold = var1 
    indx = 0

    do j = 1, ny
        do i = 1, nx 
            indx = indx + 1
            real_data(i,j) = (ichar(char_data(indx)) - 127.0) / scale + vold
            vold = real_data(i,j)
        end do
        vold = real_data(1,j)
    end do

end subroutine

subroutine check(status)

    integer, intent(in) :: status
    if(status /= nf90_noerr) then 
      stop "Check-Error"
    end if 

end subroutine

File: arl2arw.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <netcdf.h>
#include "arl2arw.h"

// Allocate data memory on heap (time(1), level, rec, lat, long)
static float data[NTIME][ZMAX][RMAX][YMAX][XMAX];

int main(int argc, char **argv)
{
    // Open ARL to file stream
    FILE *arl = fopen(argv[1], "rb");

    // Get the file size
    fseek(arl, 0, SEEK_END);
    long fsize = ftell(arl);
    rewind(arl);

    // Allocate label and ARL header array space
    char *label = malloc(sizeof(char) * LABSIZE);
    char *header = malloc(sizeof(char) * 3072); // Overflow

    // Read the standard portion of the label(50) and header(108)
    fread(label, sizeof(char), LABSIZE, arl);
    fread(header, sizeof(char), 108, arl);

    // Header INDX label
    char *hindx = varDesc(label);

    // Get grid dimensions, record length, and number of levels
    int nx = numX(header);
    int ny = numY(header);
    long nxy = nx * ny;
    long recl = nxy + LABSIZE;
    int nLvls = numLvls(header);

    free(header);

    // Allocate packed char array space
    char *cdata = malloc(sizeof(char) * ((size_t)nxy));

    int nrecs = numRecs(fsize, recl);

    // Count records per level (including 0th)
    rewind(arl);
    int current_rec = 0;
    int current_lvl = 0;
    // Store records per level
    int *rec_count = calloc(nLvls, sizeof(int));
    while (current_rec < nrecs)
    {
        ++current_rec;
        fread(label, sizeof(char), LABSIZE, arl);
        fread(cdata, sizeof(char), (size_t)nxy, arl);
        if (curLvl(label) != current_lvl)
        {
            ++current_lvl;
        }
        rec_count[current_lvl] += 1;
    }

    // NC vars ids
    int ncin_id;
    int times_var_id;
    int xlat_var_id;
    int xlong_var_id;

    // Allocate nc in var space
    char *times = malloc(sizeof(char *) * 1 * 19);
    float *xlats = malloc(sizeof(float) * 1 * ny * nx);
    float *xlongs = malloc(sizeof(float) * 1 * ny * nx);

    // Open in NC
    check(nc_open(argv[2], NC_NOWRITE, &ncin_id));

    check(nc_inq_varid(ncin_id, TIMES, &times_var_id));
    check(nc_inq_varid(ncin_id, XLAT, &xlat_var_id));
    check(nc_inq_varid(ncin_id, XLONG, &xlong_var_id));

    // times array access (time, datestrlen)
    size_t time_start[] = {0, 0};
    size_t time_count[] = {NTIME, 18};

    // grid array access (time, xlat, xlong)
    size_t grid_start[] = {0, 0, 0};
    size_t grid_count[] = {NTIME, ny, nx}; // 1 time dim! y x order!

    // Get coord and time vars
    check(nc_get_vara(ncin_id, times_var_id, time_start, time_count, times));
    check(nc_get_vara_float(ncin_id, xlat_var_id, grid_start, grid_count, xlats));
    check(nc_get_vara_float(ncin_id, xlong_var_id, grid_start, grid_count, xlongs));

    // Close in NC //
    check(nc_close(ncin_id));

    // NC dim ids
    int times_dim_id;
    int dsl_dim_id;
    int xlat_dim_id;
    int xlong_dim_id;
    int lvls_dim_id;

    int xlat_stag_dim_id;
    int xlong_stag_dim_id;
    int lvls_stag_dim_id;

    // Create NC file
    int ncout_id;
    check(nc_create(NC_OUT, NC_CLOBBER, &ncout_id));

    // Define dimensions
    check(nc_def_dim(ncout_id, TIME_NAME, NC_UNLIMITED, &times_dim_id));
    check(nc_def_dim(ncout_id, DATESTRLEN_NAME, 19, &dsl_dim_id)); // Look into

    check(nc_def_dim(ncout_id, WEST_EAST, nx, &xlong_dim_id));
    check(nc_def_dim(ncout_id, SOUTH_NORTH, ny, &xlat_dim_id));
    check(nc_def_dim(ncout_id, BOTTOM_TOP, nLvls - 1, &lvls_dim_id));

    check(nc_def_dim(ncout_id, BOTTOM_TOP_STAG, nLvls, &lvls_stag_dim_id));
    check(nc_def_dim(ncout_id, WEST_EAST_STAG, nx + 1, &xlong_stag_dim_id));
    check(nc_def_dim(ncout_id, SOUTH_NORTH_STAG, ny + 1, &xlat_stag_dim_id));

    // Time variables
    int times_dimids[] = {times_dim_id, dsl_dim_id};
    check(nc_def_var(ncout_id, TIMES, NC_CHAR, 2, times_dimids, &times_var_id));

    // Latitude longitude variables
    int grid_dimids[] = {times_dim_id, xlat_dim_id, xlong_dim_id};
    check(nc_def_var(ncout_id, XLAT, NC_FLOAT, 3, grid_dimids, &xlat_var_id));
    check(nc_def_var(ncout_id, XLONG, NC_FLOAT, 3, grid_dimids, &xlong_var_id));

    // End define mode
    check(nc_enddef(ncout_id));

    // Write NC times, lats, longs
    check(nc_put_vara(ncout_id, times_var_id, time_start, time_count, times));
    check(nc_put_vara(ncout_id, xlat_var_id, grid_start, grid_count, xlats));
    check(nc_put_vara(ncout_id, xlong_var_id, grid_start, grid_count, xlongs));
    
    free(times);
    free(xlats);
    free(xlongs);

    // Store each record var desc of every level
    char **desc[nLvls];

    // Unpak the character data
    rewind(arl);
    // For every record r in level z
    for (int z = 0; z < nLvls; ++z)
    {
        // Create string array of record count length per level
        desc[z] = malloc(sizeof(char *) * (rec_count[z]));

        for (int r = 0; r < rec_count[z]; ++r)
        {
            fread(label, sizeof(char), LABSIZE, arl);
            fread(cdata, sizeof(char), (size_t)nxy, arl);

            if (curLvl(label) != z)
            {
                fseek(arl, -(long)recl, SEEK_CUR);
                break;
            }

            // Store current record var desc
            char *cdsc = varDesc(label);
            desc[z][r] = cdsc;

            // Write label, packed character data
            if (strcmp(cdsc, hindx) != 0) // Not INDX
            {
                // Retrive label consts
                double nexp = numExp(label);
                double var1 = numVar1(label);
                // Unpack character data to 5-d data array at time 1 and level z
                unpack(nexp, var1, ny, nx, cdata, data[0][z][r]);
                // printf("%s\n", label); // DEBUG
            }
        }
    }
    
    free(label);
    free(hindx);
    free(cdata);

    // Create array of variable ids
    int varids[nrecs];
    // init current record count
    int crec = 0;

    // For every record r in each level z 
    for (int z = 0; z < nLvls; ++z)
    {
        for (int r = 0; r < rec_count[z]; ++r)
        {
            // Store current variable desc
            char *cdsc = desc[z][r];

            // Define and write surface level vars (z = 0)
            if (z == 0) 
            {
                int ndims = 3;
                int dimids[] = {times_dim_id, xlat_dim_id, xlong_dim_id};
                if ((strcmp(cdsc, "INDX") != 0))
                {
                    // Define each variable 
                    check(nc_redef(ncout_id));
                    check(nc_def_var(ncout_id, cdsc, NC_FLOAT, ndims, dimids, &varids[crec]));
                    check(nc_enddef(ncout_id));
                    
                    // Create temporary data array to reduce to conformal dimensions
                    float temp[NTIME][ny][nx];

                    // Copy data to temp array 
                    for (int j = 0; j < ny; ++j)
                    {
                        for (int i = 0; i < nx; ++i)
                        {
                            temp[0][j][i] = data[0][z][r][j][i];
                        }
                    }

                    // Create start and count vectors (time, xlats, xlongs)
                    size_t start[] = {0, 0, 0};
                    size_t count[] = {NTIME, ny, nx};
                    
                    // Put temp data array into NC
                    check(nc_put_vara_float(ncout_id, varids[crec], start, count, temp[0][0]));
                }
            }
            // TODO: Write variables at higher levels!!
            else // Define and write vars with respect to level
            {
                // Define upper level variable dimensions and dimension ids
                int ndims = 4;
                int ids[] = {times_dim_id, lvls_dim_id, xlat_dim_id, xlong_dim_id};
                
                // Ignore errors from attempted var redefines & unused gcc warning
                int stat;
                (void)stat; 

                if (strcmp(cdsc, "UWND") == 0)
                {
                    ids[3] = xlong_stag_dim_id; // Stagger long
                    stat = nc_def_var(ncout_id, cdsc, NC_FLOAT, ndims, ids, &varids[crec]);
                }
                else if (strcmp(cdsc, "VWND") == 0)
                {
                    ids[2] = xlat_stag_dim_id; // Stagger lat
                    stat = nc_def_var(ncout_id, cdsc, NC_FLOAT, ndims, ids, &varids[crec]);
                }
                else if (strcmp(cdsc, "WWND") == 0)
                {
                    ids[1] = lvls_stag_dim_id; // Stagger lvl
                    stat = nc_def_var(ncout_id, cdsc, NC_FLOAT, ndims, ids, &varids[crec]);
                }
                else // Default
                {
                    stat = nc_def_var(ncout_id, cdsc, NC_FLOAT, ndims, ids, &varids[crec]);
                }
            }
            // iterate current record count
            ++crec;
        }
    }
    
    free(rec_count);

    // Close the out NC file
    check(nc_close(ncout_id));

    // Close ARL file stream
    fclose(arl);

    return 0;
}
